### PR TITLE
ARROW-4651: [Flight] Use URIs instead of host/port pair

### PR DESCRIPTION
**Pull request by ghost**
I haven't changed the client/server construction interfaces to really take advantage of this. I would rather follow up with creating builders instead, as proposed, as that would be cleaner than special-casing the logic in the current constructors.

### CODE REVIEW

**Comment by emkornfield** [comment]
Does the change in format/ need to be voted on the mailing list?

**Comment by lihalite** [comment]
Hmm, I suppose we should, even if Flight is unstable for now.

File: format/Flight.proto
Comment (emkornfield) [inline format/Flight.proto]: I think it might be better to keep the location message and replace the internal with a single field URI.  This makes adding additional fields easier in the future if needed. Also, documenting supported protocols might be useful.

Comment (lihalite) [reply format/Flight.proto]: Yeah, makes sense.

**Comment by jacques-n** [comment]
> Does the change in format/ need to be voted on the mailing list?

I think it should be. Specifically, what is the URI format support for now. That was my main ask in general on this ticket: what do we want to support in the URI? Let's define it specifically.

**Comment by lihalite** [comment]
Ok, I will put up a proposal on the mailing list. Thanks for the comments!

**Comment by pitrou** [comment]
The protocol change proposal was formally accepted on the Arrow-dev mailing list. Now this PR needs to be rebased (and conflicts fixed) before it gets reviewed. @lihalite, do you have time for this? Otherwise, I can take up.

**Comment by lihalite** [comment]
@pitrou sorry about that, I will rebase & clean up things later today (just got back from vacation).

**Comment by lihalite** [comment]
I've now rebased this.

File: cpp/src/arrow/flight/types.h
Comment (pitrou) [inline cpp/src/arrow/flight/types.h]: Call this `ForGrpcTCP`? `ForGrpcUnix` may be added too.

File: cpp/src/arrow/flight/types.h
Comment (pitrou) [inline cpp/src/arrow/flight/types.h]: We use pointers for out-parameters, so `Location* location`. Following the Google C++ style guide: https://google.github.io/styleguide/cppguide.html#Output_Parameters

File: cpp/src/arrow/flight/types.h
Comment (pitrou) [reply cpp/src/arrow/flight/types.h]: Also, as above, should be `Location* location`.

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: This is called `unwrap`, but doesn't actually unwrap (from the name, I would expect it to return e.g. a `CAction*`).

File: cpp/src/arrow/flight/types.cc
Comment (pitrou) [inline cpp/src/arrow/flight/types.cc]: Using [std::make_shared](https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared) is generally the preferred idiom.

File: cpp/src/arrow/flight/types.cc
Comment (pitrou) [inline cpp/src/arrow/flight/types.cc]: Perhaps it can be fixed later, but this won't work for IPv6 numeric addresses, e.g. you need `grpc://[::1]:80` and not `grpc://::1:80`.

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: You already specified `ticket: Ticket` in the method signature. The only thing bad that can happen here is that `ticket` is None.

**Review by pitrou — COMMENTED** [review]
I see that the C++ server API isn't changed (`FlightServerBase::Init` takes a simple port number). Is it intentional?

**Comment by lihalite** [comment]
I intended to change the C++ server API. I think I rebased out the change on accident as I was also trying to implement the builder APIs at the same time, but now I'd rather do that in a follow up PR.

Thanks for the review, I'll get to fixing these issues!

File: cpp/src/arrow/flight/types.cc
Comment (lihalite) [reply cpp/src/arrow/flight/types.cc]: uriparser (understandably) doesn't deal with URI construction, so we'd need recognize IPv6 addresses, or create a separate method for such addresses. Or perhaps just require that the user pass `[::1]`?

Comment (pitrou) [reply cpp/src/arrow/flight/types.cc]: Yes, we can require that for now.
(uriparser seems able to deal with URI construction, but the API looks a bit terrible)

**Comment by lihalite** [comment]
While I'm at it, might as well add the builder APIs here, and make TLS-enabled services possible. In C++/Python, I did not go all the way to a Java-style builder - the changes there are much more invasive, and I don't think it's worth it until we fully settle on supporting another transport. (And even then, I think it could be done within the current APIs, or with minimal changes to them.)

**Comment by lihalite** [comment]
Rebased, tests pass.

File: cpp/src/arrow/flight/client.cc
Comment (pitrou) [inline cpp/src/arrow/flight/client.cc]: Not sure this TODO is still relevant.

File: cpp/src/arrow/util/uri.h
Comment (pitrou) [inline cpp/src/arrow/util/uri.h]: Should be `ToString` according to our conventions.

File: cpp/src/arrow/flight/types.h
Comment (pitrou) [inline cpp/src/arrow/flight/types.h]: Rather than declare friend classes, you could perhaps expose a `uri()` method?

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: I don't think we should accept `six.binary_type` here. URIs are textual.

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: You may need to add a `except ?` declaration at the end here, otherwise I don't think exceptions will be propagated to the caller.

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Shouldn't this return `CFlightDescriptor`?

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Isn's this basically `c_location = Location.unwrap(location)`?

File: python/pyarrow/tests/test_flight.py
Comment (pitrou) [inline python/pyarrow/tests/test_flight.py]: Ideally we would have a way to let gRPC bind the port and then return it (there is a race condition otherwise). But this needn't be in this PR.

File: cpp/src/arrow/flight/server.h
Comment (pitrou) [inline cpp/src/arrow/flight/server.h]: Call this `FlightServerOptions`?

**Review by pitrou — COMMENTED** [review]
Thanks for updating this. It looks mostly good to me, just some style issues and a couple other details.

File: cpp/src/arrow/flight/types.h
Comment (lihalite) [reply cpp/src/arrow/flight/types.h]: So far I've avoided that because the Uri class lives in `arrow::internal` - maybe I'm misunderstanding the intent there, though, since it is tagged with `ARROW_EXPORT`.

Comment (pitrou) [reply cpp/src/arrow/flight/types.h]: Ah, you're right. Let's keep it like that for the time being.

**Comment by lihalite** [comment]
Thanks for the feedback! I've updated things.

gRPC supports binding to port 0 for a free port, we just need a way to report that back to the API user.

**Review by pitrou — APPROVED** [review]
+1 from me on the C++ and Python changes.

**Comment by pitrou** [comment]
Does this need reviewing on the Java side?

**Comment by lihalite** [comment]
Thanks @pitrou. For the Java side, perhaps @jacques-n could take a look again?

**Comment by lihalite** [comment]
Updated to fix an inadvertent API breakage. (`Ticket` is quite useless if you can't read its value...)

**Comment by lihalite** [comment]
There are actually lots of little things I'm noticing here now that I'm trying to test internally, so please hold off while I fix things over the next couple days...apologies for the trouble.

**Comment by lihalite** [comment]
Ok, this should be ready now; fixed some inadvertent API breakages.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightEndpoint.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightEndpoint.java]: Why the move away from streams in some of these changes?

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: Can we include support for domain socket in this patch? It would be nice to confirm that they also work cross platform.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: How about we make the executor configurable as part of this builder? Same with max message size below.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: Can't we have a default for certChain? Also, does it make sense to require it be a file?

File: java/flight/src/main/java/org/apache/arrow/flight/Location.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/Location.java]: if this is public, we should validate the provided uri to make sure it is supported.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: these properties should be settable via builder methods (rather than just constructor)

File: java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java]: let's add a private constructor and final class so people don't create this

File: java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java]: dumb question, what is grpc verus grpc+tcp?

File: python/pyarrow/tests/test_flight.py
Comment (pitrou) [inline python/pyarrow/tests/test_flight.py]: `pa.ArrowInvalid` or `ValueError` would be more precise?

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Writing `tls_cert_chain=None, tls_private_key=None` instead will make the signature more informative (and also avoid typos in argument names).

**Review by pitrou — COMMENTED** [review]
Looks mostly good to me on the C++/Python side. Just a few nits.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: For gRPC/Netty, they are Unix-only and require a Unix-only dependency. If that's OK, I'll add tests for that.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: Sure - I was going to do that in a followup, but I'll include that here.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: I'll add an overload for InputStream.

We may be able to hard-code some platform-specific paths to try.

File: java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/LocationSchemes.java]: Not a dumb question :) So far, they're aliases for each other (so the default "grpc" protocol is insecure gRPC over TCP)

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jacques-n) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: Let's add this and only fail if we're running some place else.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightEndpoint.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightEndpoint.java]: Ah, here it's because I wanted the checked exception to be propagated - the stream hides this.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java]: Oh wait, I mixed this up. This is the server, so you must provide the certificate chain and key (it's the cert/key the server presents). The client defaults to a platform-specific value already, though we should let the client specify a certificate store to check against.

**Comment by wesm** [comment]
It looks like this is near merge-readiness. @jacques-n can you review/sign off on the Java changes?

**Comment by lihalite** [comment]
Rebased with master as well.

**Comment by wesm** [comment]
I think Jacques is traveling right now so it may be a little time before we can get a go-ahead from the Java side. Would my review on the C++/Python side be helpful?

**Comment by lihalite** [comment]
I'd appreciate any feedback! On the Java side, I'm not in a particular rush to get this merged, and I can keep up with master, I'd just like to make sure I get everything in for 0.14.

**Comment by wesm** [comment]
Cool, BTW I'm guessing on a somewhat longer release timeline than usual for 0.14 to give various in-flight efforts to sort themselves out, e.g. toward end of June / beginning July

**Comment by emkornfield** [comment]
Please make sure to run rebase any java CLs and re-run CI to make sure javadoc's are in place.

**Comment by wesm** [comment]
@jacques-n I think you are the last approver on this PR

**Comment by jacques-n** [comment]
Looks good to me. Thanks for pulling this together @lihalite!

**Comment by wesm** [comment]
thanks @lihalite! It might have already been discussed here, but what is the testing strategy for TLS-enabled Flight going to be? Unless I missed it is doesn't seem that this is tested now, can we open a JIRA?

**Comment by lihalite** [comment]
It's not currently tested, JIRA: https://jira.apache.org/jira/browse/ARROW-5397

We'll need some way to generate certs/keys.

**Comment by pitrou** [comment]
The easiest thing to do is to store self-signed certs and the corresponding private key in the repo.
Preferable along with the script to regenerate them when the expiry date is reached.

**Pull request closed without merging**
